More Structure in Error Object (#718)
SAP/cloud-sdk-js: knowledge-base/adr/structure-of-error-object.md

This becomes even more problematic since there are no `throws declaration` possible.

### Option B: Include the Root cause error Message 

We could at least include  a `root` property to the error:

```typescript
interface ErrorWithRoot {
    name: string;
    message: string;
    stack?: string;
    rootMessage?:string   
}
```

the variable `rootMessage` would contain `Request failed with status code 400` in the example above.

Pros:
- You can access the root message  quickly
- No inheritance

Cons:
- No structured information on the kind of error
- A check if the property is there is needed

### Option C: Make the RequestBuilder non-throwy

In the API discussion [ADR](https://github.com/SAP/cloud-sdk-js/pull/709) we discussed a change to include more request and response data:

```typescript
const [buPa, req, res]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().getAll().executeRaw(destination);
```

One could include a `httpNoThrow` flag to the `execute` methods. 
If switched on the request builder will not throw HTTP related error but include them in the return. 

```typescript

> const [buPa, httpError]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().httpNoThrow().getAll().execute(destination);

Reviewer: Maybe adjust the type? `const [buPa, httpError]: [BusinessPartner, Request, Response]`

# More Structure in Error Object

## Status

> proposed

Reviewer 2: decided

    at Object.errorWithCause (XXX/srv/node_modules/@sap-cloud-sdk/util/dist/error.js:5:20)
    at constructError (XXX/srv/node_modules/@sap-cloud-sdk/core/dist/odata/common/request/odata-request.js:295:23)
    at XXX/srv/node_modules/@sap-cloud-sdk/core/dist/odata/common/request/odata-request.js:265:51
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at ...
Caused by:
Error: Request failed with status code 400
    at createError (XXX/srv/node_modules/axios/lib/core/createError.js:16:15)
    at settle (XXX/srv/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (XXX/srv/node_modules/axios/lib/adapters/http.js:236:11)
    at IncomingMessage.emit (events.js:326:22)
    at endReadableNT (_stream_readable.js:1223:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
```

but it is totally unstructured and a pure string. 
The issue raised by the customer asked to have the root cause in a more structured way like:

```typescript
interface HttpError {
    name: string;
    statucCode:number
    message: string;    
}
```
This ADR decides the way to improve the situation for the customer.

## Options

> ### Option A: Extending the Error Interface/Object

Reviewer 2: ```suggestion
### Option A: Extending the Error interface/object
```

We could extend the Error object: 

```typescript
class HttpError extends Error{
  constructor(...args) {
    super(...args); 
    this.name = "HttpError"; 
  }
}
```
and return a more concrete Error in the `errorWithCause` method.
There is a `isAxiosError` to detect HTTP errors [see here](https://github.com/axios/axios/pull/1419)

Pros:
- Typed information specific to error: Http, IO, ...
- Detailed information like HttpStatus etc can be accessed easily.

Cons:
- Not very JavaScripty
- We need to find the right type in the `errorWithCause` -> big exception mapper factory
- The consumer needs some kind of "down cast" to find the actual error type returned (catch blocks, instanceof check)
This becomes even more problematic since there are no `throws declaration` possible.

> ### Option B: Include the Root cause error Message 

Reviewer 2: ```suggestion
### Option B: Include the original error
```

We could at least include  a `root` property to the error:

```typescript
class ErrorWithRoot extends Error {
    name: string;
    message: string;
    stack?: string;

>     originalError?:Error   
>     get rootError():Error|undefined{}
>     //recursibly check if originalError is of type ErrorWithRoot and return the first non  ErrorWithRoot error in this chain.

Reviewer 2: ```suggestion
    originalError?: Error;
    get rootError(): Error | undefined {
      //recursibly check if originalError is of type ErrorWithRoot and return the first non  ErrorWithRoot error in this chain.
    }
```

}
```

> the variable `rootMessage` would contain `Request failed with status code 400` in the example above.

Reviewer 2: Please update this sentence ;)

Pros:
- You can access the root message  quickly
- No inheritance

Cons:
- No structured information on the kind of error
- A check if the property is there is needed

### Option C: Make the RequestBuilder non-throwy

In the API discussion [ADR](https://github.com/SAP/cloud-sdk-js/pull/709) we discussed a change to include more request and response data:

```typescript
const [buPa, req, res]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().getAll().executeRaw(destination);
```

One could include a `httpNoThrow` flag to the `execute` methods. 
If switched on the request builder will not throw HTTP related error but include them in the return. 

```typescript
const [buPa, httpError]: [BusinessPartner, HttpError] = await BusinessPartner.requestBuilder().httpNoThrow().getAll().execute(destination);
```

The `httpError` object contains information on  HTTP errors appearing during the request. 

Pros:
- Information on Http Errors typed
- No casting needed 

Cons:
- Errors are handled differently (Http vs. deserializing errors for example).

### Option D: Introduce `failable` return types

> A pattern I found [here](https://medium.com/@dhruvrajvanshi/making-exceptions-type-safe-in-typescript-c4d200ee78e9) is a failable type:

Reviewer 2: I would avoid "I" and personal opinions in an ADR, I think that is irrelevant.
```suggestion
This pattern](https://medium.com/@dhruvrajvanshi/making-exceptions-type-safe-in-typescript-c4d200ee78e9) is a failable type:
```

```typescript
type Failable<ResultType, ErrorType=Error> = {
  isError: true;
  error: ErrorType;
} | {
  isError: false;
  value: ResultType;
}
``` 

Introducing such a types is a bit like a throw declaration and you can list the exception which are appearing.

```typescript
function execute():Failable<BusinessPartner,AxiosError| ParseError | Error>
```

> Nothe that his also goes a bit in the direction of checked exceptions but you have union types and can define type aliases which avoid that you have to change a zillion files you add one more error cause.

Reviewer 2: ```suggestion
Note that his also goes a bit in the direction of checked exceptions but you have union types and can define type aliases which avoid that you have to change a zillion files you add one more error cause.
```

Pros:
- Error handling strongly typed
- Errors are clearly handled by return not catch

Cons:
- People are not used to this pattern but expect to catch stuff
- The customer needs to check the possible types (down cast problem from above)
- Breaking API change

## Decision

We take option B.
Check convention on MDN how what to throw i.e. should you extend from the error object. 
Also adjust the name `ErrorWithRoot` to something better.

The colleagues from exchange rate lib had problems with the [prototype chain](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget).
Just consider these kind of problems.

## Consequences

> none

Reviewer 2: I guess this needs an update ;)
